### PR TITLE
Feat(eos_designs): Interface "LACP Fallback Individual" Support

### LINKED ISSUE #3367: Interface "LACP Fallback Individual" Support
### Enhancement summary

It is not possible right now to configure "LACP Fallback Individual" via the AVD data model, we only support the "static" option currently.  This is a common requirement for automated server deployments in order to bootstrap bare-metal servers via a different VLAN or interface type (for example) than is configured against the port-channel interface.

EOS feature functionality is described here
https://arista.my.site.com/AristaCommunity/s/article/configuring-lacp-fallback-individual-ports-on-arista-switches

### Which component of AVD is impacted

eos_designs

### Use case example

This is a common requirement for automated server deployments in order to bootstrap bare-metal servers via a different VLAN then is configured against the port-channel interface.

### Describe the solution you would like

The ability to specify different interface configuration parameters at the member interface level than are specified at the port-channel level 

### Describe alternatives you have considered

There is no alternative in EOS, we are currently using the structured_configuration parameter as a workaround to allow for deployment to continue

### Additional context

_No response_

### Contributing Guide

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by bjmeuer**
## Change Summary

Adding "LACP Fallback Individual" to the eos_designs.  This is a common requirement for automated server deployments in order to bootstrap bare-metal servers via a different VLAN or interface type (for example) than is configured against the port-channel interface. Also adding the ability to specify different interface configuration parameters at the member interface level than are specified at the port-channel level by using port_profiles.

## Related Issue(s)

Fixes #3367 

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes
The option "individual" was added to the "mode" key of the lacp_fallback setting. Additionally added the setting "individual" with the key "profile" where you can reference a port_profile from which you want to inherit configuration for the member interfaces. The following config from a port_profile will be inherited: mode, vlans, native_vlan, native_vlan_tag.

```
servers:
  - name: server1
    adapters:
    - switch_ports: ['Ethernet10/1', 'Ethernet10/1']
      ...
        lacp_fallback: 
          mode: individual
          individual:
            profile: ESX_TRUNK
```

## How to test
From the molecule tests:

```
servers:
  # port-channel with lacp fallback individual
  - name: INDIVIDUAL_1
    adapters:
      - switches: [connected_endpoints, connected_endpoints]
        switch_ports: [Ethernet12, Ethernet13]
        port_channel:
          mode: "active"
          description: "INDIVIDUAL_1"
          lacp_fallback:
            mode: individual
  # port-channel with lacp fallback individual and profile for member interfaces
  - name: INDIVIDUAL_2
    adapters:
      - switches: [connected_endpoints, connected_endpoints]
        switch_ports: [Ethernet14, Ethernet15]
        port_channel:
          mode: "active"
          description: "INDIVIDUAL_2"
          lacp_fallback:
            mode: individual
            individual:
              profile: INDIVIDUAL_TRUNK

port_profiles:
  - profile: INDIVIDUAL_TRUNK
    mode: trunk
    vlans: "1,2,3,4,5,6,7,123,234"
    native_vlan_tag: False
    native_vlan: 123

```

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/connected_endpoints.cfg [hunk 1]
```diff
@@ -67,6 +81,38 @@ interface Ethernet11
...
+!
+interface Ethernet14
+   description INDIVIDUAL_2
+   no shutdown
+   switchport trunk native vlan 123
+   switchport trunk allowed vlan 1,2,3,4,5,6,7,123,234
+   switchport mode trunk
+   switchport
+   channel-group 14 mode active
+   lacp port-priority 8192
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/connected_endpoints.cfg]: The port-priority should not be needed for individual mode.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/connected_endpoints.cfg [hunk 2]
```diff
@@ -67,6 +81,38 @@ interface Ethernet11
    description PHYSICAL_PORT_DESCRIPTION_2
    no shutdown
    channel-group 10 mode active
+!
+interface Ethernet12
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/connected_endpoints.cfg:78]: This config will not work. So I think we need to require the individual profile to be set, if lacp fallback mode is "individual"

File: ansible_collections/arista/avd/roles/eos_designs/docs/tables/port-profiles.md [hunk 3]
```diff
@@ -434,8 +436,15 @@
...
-            # Currently only static mode is supported.
-            mode: <str; "static">
+            # Either static or individual mode is supported.
+            mode: <str; "static" | "individual">
+
+            # Define parameters for port-channel member interfaces. Applies only if LACP fallback is set to "individual".
+            individual:
+
+              # Port-profile name to inherit configuration.
+              # Currently the "mode", "vlans", "native_vlan" and "native_vlan_tag" from the profile are supported to be inherited.
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/docs/tables/port-profiles.md]: Why do we need to limit this?

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py [hunk 4]
```diff
@@ -139,8 +139,23 @@ def _get_ethernet_interface_cfg(self, adapter: dict, node_index: int, connected_
...
+                            "The 'profile' of every port-channel lacp fallback individual setting must be defined in the 'port_profiles'. First occurence seen"
+                            f" of a missing profile is '{get(adapter, 'port_channel.lacp_fallback.individual.profile')}' for the connected endpoint with the"
+                            f" name '{connected_endpoint['name']}'."
+                        )
+
+                    ethernet_interface["type"] = "switched"
+                    ethernet_interface["mode"] = profile.get("mode")
+                    ethernet_interface["vlans"] = profile.get("vlans")
+                    ethernet_interface["native_vlan_tag"] = profile.get("native_vlan_tag")
+                    ethernet_interface["native_vlan"] = profile.get("native_vlan")
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py]: I think we may have to split out the config from the "NOT a port-channel member" section below so we can have call a common function for individual or non-member ports. (The function should take the adapter for non-member ports and the profile for indivual.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/connected_endpoints.cfg (hunk 1 above)
Comment (bjmeuer) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/connected_endpoints.cfg]: removed for the individual mode

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/connected_endpoints.cfg (hunk 2 above)
Comment (bjmeuer) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/connected_endpoints.cfg:78]: added it to the documentation and also put throw a missing variable error if the profile is missing when individual mode is used. Added a negative test as well.

File: ansible_collections/arista/avd/roles/eos_designs/docs/tables/port-profiles.md (hunk 3 above)
Comment (bjmeuer) [reply ansible_collections/arista/avd/roles/eos_designs/docs/tables/port-profiles.md]: removed limitation

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py (hunk 4 above)
Comment (bjmeuer) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py]: created a shared function which can be used by both

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py [hunk 5]
```diff
@@ -78,6 +78,36 @@ def ethernet_interfaces(self) -> list | None:
 
         return None
 
+    def _update_ethernet_interface_cfg(self, adapter: dict, ethernet_interface: dict, connected_endpoint: dict, short_esi: str, node_index: int) -> dict:
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py]: This change is great. We just need to move a few of the items back under the non-portchannel case. From what I can figure out I would move poe, evpn_ethernet_segment and link_tracking_groups, so these do not apply for the fallback case.
I don't know about l2_mtu and l2_mru. Basically we need to remove anything that could potentially affect the associated port-channel when it is in non-fallback mode.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py [hunk 6]
```diff
@@ -139,22 +169,26 @@ def _get_ethernet_interface_cfg(self, adapter: dict, node_index: int, connected_
...
+
+            if get(adapter, "port_channel.lacp_fallback.mode") == "individual":
+                # if fallback is set to individual a profile has to be defined
+                if get(adapter, "port_channel.lacp_fallback.individual") is None or get(adapter, "port_channel.lacp_fallback.individual.profile") is None:
+                    raise AristaAvdMissingVariableError(
+                        "A Port-channel which is set to lacp fallback mode 'individual' must have a 'profile' defined. Profile definition is missing for"
+                        f" the connected endpoint with the name '{connected_endpoint['name']}'."
+                    )
+
+                # check if the referred name exists in the global evpn_vlan_bundles
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py]:
```suggestion
                # Verify that the referred profile exists under port_profiles
```

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py [hunk 7]
```diff
@@ -139,22 +169,26 @@ def _get_ethernet_interface_cfg(self, adapter: dict, node_index: int, connected_
...
-
-                    ethernet_interface["type"] = "switched"
-                    ethernet_interface["mode"] = profile.get("mode")
-                    ethernet_interface["vlans"] = profile.get("vlans")
-                    ethernet_interface["native_vlan_tag"] = profile.get("native_vlan_tag")
-                    ethernet_interface["native_vlan"] = profile.get("native_vlan")
+
+            if get(adapter, "port_channel.lacp_fallback.mode") == "individual":
+                # if fallback is set to individual a profile has to be defined
+                if get(adapter, "port_channel.lacp_fallback.individual") is None or get(adapter, "port_channel.lacp_fallback.individual.profile") is None:
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py]:
```suggestion
                if (profile_name := get(adapter, "port_channel.lacp_fallback.individual.profile")) is None:
```

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py [hunk 8]
```diff
@@ -139,22 +169,26 @@ def _get_ethernet_interface_cfg(self, adapter: dict, node_index: int, connected_
...
+                    raise AristaAvdMissingVariableError(
+                        "A Port-channel which is set to lacp fallback mode 'individual' must have a 'profile' defined. Profile definition is missing for"
+                        f" the connected endpoint with the name '{connected_endpoint['name']}'."
+                    )
+
+                # check if the referred name exists in the global evpn_vlan_bundles
+                if (profile := get_item(self._hostvars["port_profiles"], "profile", get(adapter, "port_channel.lacp_fallback.individual.profile"))) is None:
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py]:
```suggestion
                if not (profile := self.shared_utils.get_merged_profile(profile_name)):
```

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py [hunk 9]
```diff
@@ -141,6 +171,25 @@ def _get_ethernet_interface_cfg(self, adapter: dict, node_index: int, connected_
             )
             if get(adapter, "port_channel.lacp_fallback.mode") == "static":
                 ethernet_interface["lacp_port_priority"] = 8192 if node_index == 0 else 32768
+
+            if get(adapter, "port_channel.lacp_fallback.mode") == "individual":
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py]:
```suggestion
            elif get(adapter, "port_channel.lacp_fallback.mode") == "individual":
```

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py (hunk 5 above)
Comment (bjmeuer) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py]: I think we can also move ptp back as this has normally no relevance the fallback case. I will test the l2 configs if they have an impact on the port-channel

Comment (bjmeuer) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/ethernet_interfaces.py]: did move some settings back after testing

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/port_channel_interfaces.py [hunk 10]
```diff
@@ -175,7 +175,10 @@ def _get_port_channel_interface_cfg(self, adapter: dict, port_channel_interface_
             port_channel_interface["mlag"] = channel_group_id
 
         # LACP Fallback
-        if port_channel_mode in ["active", "passive"] and (lacp_fallback_mode := get(adapter, "port_channel.lacp_fallback.mode")) == "static":
+        if port_channel_mode in ["active", "passive"] and (
+            (lacp_fallback_mode := get(adapter, "port_channel.lacp_fallback.mode")) == "static"
+            or (lacp_fallback_mode := get(adapter, "port_channel.lacp_fallback.mode")) == "individual"
+        ):
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/port_channel_interfaces.py]:
```suggestion
        if port_channel_mode in ["active", "passive"] and             (lacp_fallback_mode := get(adapter, "port_channel.lacp_fallback.mode")) in ["static", "individual"]:
```

**Review by ClausHolbechArista — COMMENTED** [review]
Minor code optimization. Done by hand so may require linting.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/port_channel_interfaces.py (hunk 10 above)
Comment (bjmeuer) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/port_channel_interfaces.py]: like that one

**Review by ClausHolbechArista — APPROVED** [review]

**Review by ankudinov — APPROVED** [review]
LGTM. Tested various combinations in Codespace

**Pull request merged**